extensions/amp-analytics/0.1/activity-impl.js

```diff
@@ -87,7 +87,13 @@ class ActivityHistory {
       this.prevActivityEvent_ = activityEvent;
     }
 
-    if (this.prevActivityEvent_.time < activityEvent.time) {
+    if (
+      this.prevActivityEvent_.time < activityEvent.time ||
+      // when pushed event with "inactive" type from (mouseleave)
+      // prevActivityEvent has event with same time but "active" type from (mousemove)
+      (this.prevActivityEvent_.time === activityEvent.time &&
```

Comment:
You don't need this, `findEngagedTimeBetween` already handles `INACTIVE` event and case when the two events time are the same.

Reply:
Let's talk about this.
As i understand `findEngagedTimeBetween(activityEvent, time)` return diff between last pushed activity and time; max return value 5 for active events, and 0 for inactive;

If we leave only `this.prevActivityEvent_.type !== activityEvent.type` check, 
new activities with same type but different time won't be pushed;
and when we call 
`getTotalEngagedTime() {
    const secondsSinceStart = Math.floor(this.getTimeSinceStart_() / 1000);
    return this.activityHistory_.getTotalEngagedTime(secondsSinceStart);
  }`
we can get maximum +5 from first acive event (if we have no inactive events)

Reply:
@zhouyx did I understand you correctly?

Reply:
@Galphimbl maybe you can create a test case that demonstrates why this check is necessary?

Reply:
All you need is the `if (this.prevActivityEvent.time < activityEvent.time)`, even that check is only for extra protection because prevActivityEvent time should always be less than the next activityEvent. 
https://github.com/ampproject/amphtml/blob/a0f855807981b65fa662395e1ae30f59c68be3fa/extensions/amp-analytics/0.1/activity-impl.js#L91-L94

https://github.com/ampproject/amphtml/blob/a0f855807981b65fa662395e1ae30f59c68be3fa/extensions/amp-analytics/0.1/activity-impl.js#L56-L64

engagementBonus is set to 0 unless the event type is INACTIVE. So four cases

1. `previsouActivityEvent_.type = INACTIVE` and `activityEvent_.type = INACTIVE`
`engagedTimeBetween` would be Math.min(timeDiff, 0) = 0 

2. `previsouActivityEvent_.type = ACTIVE` and `activityEvent_.type = INACTIVE`
`engagedTimeBetween` would be Math.min(timeDiff, 5)

3 . `previsouActivityEvent_.type = INACTIVE` and `activityEvent_.type = ACTIVE`
`engagedTimeBetween` would be Math.min(timeDiff, 0) = 0 

4. `previsouActivityEvent_.type = ACTIVE` and `activityEvent_.type = INACTIVE`
`engagedTimeBetween` would be Math.min(timeDiff, 5)

I don't see an issue without the check.

Reply:
@zhouyx that's right, but issue little different. 
When user leave mouse from the document, browser dispatch `mousemove // ACTIVE`  event and moment later `mouseleave // INACTIVE`.
As in `ActivityEvent.time`  value in seconds, when we push last `mouseleave // INACTIVE` event, `previousActivityEvent_`  has `mousemove // ACTIVE` event with same time -> `this.prevActivityEvent.time < activityEvent.time` return `false` -> 
`prevActivityEvent` stay on `ACTIVE` -> 
`getTotalEngagedTime()` return value up to `prevActivityEvent.time + 5`.
expected result `getTotalEngagedTime()` return `prevActivityEvent.time` becouse user leave mouse from the document.

for less code we can change `this.prevActivityEvent.time < activityEvent.time` to `this.prevActivityEvent.time <= activityEvent.time`

later i write tests to show this case.

Reply:
I think we can do the following to make code cleaner.
```
push(activityEvent) {
  if (this.prevActivityEvent_ && this.prevActivityEvent_.time < activityEvent.time) {
    this.totalEngagedTime_ +=...
  }
  this.prevActivityEvent_ = activityEvent;
}
```

It's also fine to remove the `this.prevActivityEvent_.time < activityEvent.time` check